### PR TITLE
[OUDS] feat: Add 'Borders' token, utilities and documentation

### LINKED ISSUE #2582: [OUDS] Tokens: Border
## Description

The aim of this issue is to study and implement the border tokens taking into account the cohesive multi-platform approach, the customization by libraries inheriting OUDS Web, but also what we're using for OUDS Web: Boosted, Bootstrap etc.
- Define the possibilities
- Define the constraints
- Define the impacts
- Implement the tokens (with a future Style Dictionary usage in mind)

Reminder: there will be in the end 3 layers of tokens:
- **Raw primitive values: Insertion of brand foundations.**
  - These tokens will be customizable by libraries inheriting OUDS Web.
  - These tokens cannot be used by developers using OUDS Web.
- **Semantic applications: Depends on raw primitive values.**
  - Mapping between these tokens and the raw primitive values will be customizable by libraries inheriting OUDS Web.
  - These tokens can be used by developers using OUDS Web.
- **Component applications: Depends on semantic applications.**:
  - Mapping between these tokens and the semantic tokens should rarely be customized by libraries inheriting OUDS Web.
  - These tokens can be used by developers using OUDS Web.

### Raw primitive values

|  Token name | Token value |
|---|---|
| border-base  | 4  |

#### Width

|  Token name | Token value |
|---|---|
| border-width-0  | {border-base}*0 = 0  |
| border-width-25  | {border-base}*0.25 = 1  |
| border-width-50  | {border-base}*0.5 = 2  |
| border-width-75  | {border-base}*0.75 = 3  |
| border-width-100 | {border-base}*1 = 4  |
| border-width-150  | {border-base}*1.5 = 6  |
| border-width-200  | {border-base}*2 = 8  |

#### Radius

|  Token name | Token value |
|---|---|
| border-radius-0  | {border-base}*0 = 0  |
| border-radius-25  | {border-base}*0.25 = 1  |
| border-radius-50  | {border-base}*0.5 = 2  |
| border-radius-75  | {border-base}*0.75 = 3  |
| border-radius-100  | {border-base}*1 = 4  |
| border-radius-150  | {border-base}*1.5 = 6  |
| border-radius-200  | {border-base}*2 = 8  |
| border-radius-300  | {border-base}*3 = 12  |
| border-radius-400  | {border-base}*4 = 16  |
| border-radius-500  | {border-base}*5 = 20  |
| border-radius-600  | {border-base}*6 = 24  |
| border-radius-800  | {border-base}*8 = 32  |
| border-radius-9999  | 2000  |

#### Style

|  Token name | Token value |
|---|---|
| border-style-none  | none  |
| border-style-solid  | solid  |
| border-style-dashed  | dashed  |
| border-style-dotted  | dotted  |

#### Width

|  Token name | Token value |
|---|---|
| border-width-none | border-width-0 |
| border-width-default | border-width-25 |
| border-width-thin | border-width-25 |
| border-width-thick | border-width-50 |
| border-width-thicker | border-width-75 |
| border-width-thickest | border-width-100 |
| border-width-interactive-primary-focus | border-width-100 |

#### Radius

|  Token name | Token value |
|---|---|
| border-radius-none | border-radius-0 |
| border-radius-default | border-radius-0 |
| border-radius-short | border-radius-75 |
| border-radius-medium | border-radius-150 |
| border-radius-tall | border-radius-300 |
| border-radius-pill | border-radius-9999 |
| border-radius-circle | 50% |

#### Style

|  Token name | Token value |
|---|---|
| border-style-default | border-style-solid |
| border-style-drag | border-style-dashed |

--- 

## Study

- [x] Study the architecture of these tokens, and define the technical details

##### Border opacity

In Bootstrap, there's a utility set for [border opacity](https://getbootstrap.com/docs/5.3/utilities/borders/#opacity) such as `.border-opacity-25`, `.border-opacity-75`.
* As it's done for the `.opacity` utilities (see https://github.com/Orange-OpenSource/Orange-Boosted-Bootstrap/issues/2575), this part must be hidden in the documentation
* The question was whether we need to provide new ones like `.border-opacity-transparent`, `.border-opacity-weak`, etc. to match the `.opacity-transparent`, `.opacity-weak`. Based on feedback from @MaximeTonnerre, using color semantic tokens in Figma lock the possibility to play with the opacity. Since designers won't create dedicated border color opacity tokens, the rule is that our interfaces shouldn't have any semi-transparent borders. So there's no need to create such new utilities.
  * Except maybe the "extreme" ones for "transparent" and "opaque"?
  * If there are no border opacity utilities, let's keep the "Opacity" section in the documentation with a paragraph explaining that developers/interfaces shouldn't play with border opacity to respect OUDS.

### PR DESCRIPTION
**Pull request by vprothais**
### Related issues

Closes #2582
Linked to #2589

### Description

Add Border utilities based on design tokens with corresponding documentation.

LM:
- Kept the change of `.border-pill` and `.border-circle` in the doc but should surely be backported in Bootstrap as well.
- Should we introduce parts of the doc with a tutorial to see the semantic variables in action in components tokens later on?

|  Branch  |  `ouds/main`  |  `ouds/main-vp-border-token`  | Inflate/deflate |
|---|---|---|---|
|  `ouds-web.css` size  |  360 607  |  357 712  |  -1% (-2895 B)  |
|  `ouds-web-bootstrap` size  |  360 835  |  365 337  |  +1% (+4502 B)  |
|  `ouds-web-grid.css` size  |  73 461  |  73 461  |  0% (0 B)  |
|  `ouds-web-reboot.css` size  |  19 357  |  19 047  |  -2% (-310 B)  |
|  `ouds-web-utilities.css` size  |  117 910  |  115 153  |  -2% (-2757 B)  |

### Todo
- [x] Finalize documentation concerning border opacity (see description in #2582)
- [x] Finalize documentation concerning variables, sass variables and maps
- [x] Compare documentation with Bootstrap version (current version is based on Boosted one)
- [x] Add content to both migration documentation
- [x] Add unit test

#### To add to #2589

- [ ] Uncomment elements whenever colors (and so border colors) will be reintroduced.

### Types of change

- New feature (non-breaking change which adds functionality)

### Live previews

- <https://deploy-preview-2687--boosted.netlify.app/docs/0.0/utilities/borders/>

#### Contribution

- [x] I have read the [contributing guidelines](https://github.com/Orange-OpenSource/Orange-Boosted-Bootstrap/blob/main/.github/CONTRIBUTING.md)

#### Accessibility

- [x] My change follows accessibility good practices; I have at least run axe

#### Design

- [x] My change respects the design guidelines defined in [Orange Design System](https://oran.ge/dsweb)
- [x] My change is compatible with a responsive display

#### Development

- [x] My change follows the [developer guide](https://github.com/Orange-OpenSource/Orange-Boosted-Bootstrap/wiki/Developer-guide)
- [x] I have added JavaScript unit tests to cover my changes
- [x] I have added SCSS unit tests to cover my changes

#### Documentation

- [x] My change introduces changes to the documentation and/or I have updated the documentation accordingly

### Checklist (for Core Team only)

- [x] My change introduces changes to the migration guide
- [x] My new component is well displayed in [Storybook](https://deploy-preview-2687--boosted.netlify.app/storybook)
- [x] My new component is compatible with RTL
- [x] Manually run [BrowserStack tests](https://github.com/Orange-OpenSource/Orange-Boosted-Bootstrap/actions/workflows/browserstack.yml)
- [x] Manually test browser compatibility with BrowserStack (Chrome >= 60, Firefox >= 60 (+ ESR), Edge, Safari >= 12, iOS Safari, Chrome & Firefox on Android)
- [ ] Code review
- [ ] Design review
- [ ] A11y review

#### After the merge

- [ ] Manually launch [Percy tests](https://github.com/Orange-OpenSource/Orange-Boosted-Bootstrap/actions/workflows/percy.yml)

### CODE REVIEW

File: site/content/docs/0.0/utilities/borders.md [hunk 1]
```diff
@@ -8,4 +8,293 @@ aliases:
...
+{{< /example >}}
+
+{{< bootstrap-compatibility >}}
+
+Use the scaling classes for larger or smaller rounded corners. Sizes range from `0` to `5`, and can be configured by modifying the utilities API.
+
+{{< example class="bd-example-rounded-utils" >}}
+{{< placeholder width="75" height="75" class="rounded-0" title="Example non-rounded image" color="#999" >}}
+{{< placeholder width="75" height="75" class="rounded-1" title="Example small rounded image" color="#999" >}}
+{{< placeholder width="75" height="75" class="rounded-2" title="Example default rounded image" color="#999" >}}
```
Comment (vprothais) [inline site/content/docs/0.0/utilities/borders.md:207]: this utility is a bit weird, the map is like that :
```scss
        1: var(--#{$prefix}border-radius-sm),
        2: var(--#{$prefix}border-radius),
        3: var(--#{$prefix}border-radius-lg),
```
In OUDS, the variable "border-radius" is equal to the token "default", valued to 0. But "radius sm" is mapped to "radius-short". So we ended up with `rounded-2` that doesn't have any border radius but `rounded-1` has.
We probably need to change the mapping in utilities .

File: site/content/docs/0.0/utilities/borders.md [hunk 2]
```diff
@@ -8,4 +8,293 @@ aliases:
...
+<span class="border border-5"></span>
+{{< /example >}}
+
+{{< /bootstrap-compatibility >}}
+
+## Radius
+
+Add classes to an element to easily round its corners.
+
+<!-- TODO : This example with default rounded value doesn't have any sens since the default value is 0. Do we keeo it ? -->
```
Comment (vprothais) [inline site/content/docs/0.0/utilities/borders.md:168]: since `default`is set to 0, the `rounded` class doesn't add any border-radius. This also applies to `rounded-top` and such. In another brand this could be usefull but not in OUDS. Do we drop this or explain that border-radius are not allowed by default. This default management may need to be reviewed since it also generate a CSS variable.

File: site/content/docs/0.0/utilities/borders.md (hunk 1 above)
Comment (louismaximepiton) [reply site/content/docs/0.0/utilities/borders.md:222]: This is a bit weird, but I'm wondering if we should keep the CSS variables for the Bootstrap compatibility only or not. If yes, maybe it's not that weird to have this kind of thing.

File: scss/tokens/_raw.scss [hunk 3]
```diff
@@ -16,4 +16,36 @@ $ouds-opacity-500: .32 !default;
...
+$ouds-border-radius-75:   $ouds-border-base * .75 !default;
+// $ouds-border-radius-100:  $ouds-border-base !default;
+$ouds-border-radius-150:  $ouds-border-base * 1.5 !default;
+// $ouds-border-radius-200:  $ouds-border-base * 2 !default;
+$ouds-border-radius-300:  $ouds-border-base * 3 !default;
+// $ouds-border-radius-400:  $ouds-border-base * 4 !default;
+// $ouds-border-radius-500:  $ouds-border-base * 5 !default;
+// $ouds-border-radius-600:  $ouds-border-base * 6 !default;
+// $ouds-border-radius-800:  $ouds-border-base * 8 !default;
+$ouds-border-radius-9999: 2000px !default;
```
Comment (julien-deramond) [inline scss/tokens/_raw.scss:45]: This one (pill) will be dropped from the tokens as said today during the workshop.
It can be handled like `"circle": 50%` in `$ouds-border-radiuses`.

File: .stylelintignore [hunk 4]
```diff
@@ -1,4 +1,5 @@
 **/*.min.css
+**/*.test.scss
```
Comment (louismaximepiton) [inline .stylelintignore:2]: Maybe a better way could be to disable stylelint inside the file only. I don't have any strong opinion on this subject but felt fine to ignore tests .

File: site/content/docs/0.0/utilities/borders.md [hunk 5]
```diff
@@ -8,4 +8,311 @@ aliases:
...
+
+<div class="p-3 bg-info bg-opacity-10 border border-info border-start-0">
+  Changing border color and width
+</div>
+{{< /example >}}
+</details>-->
+
+## Opacity
+
+{{< callout danger >}}
```
Comment (louismaximepiton) [inline site/content/docs/0.0/utilities/borders.md:107]: Just a quick question on this, do we really need this callout since the remaining is hidden by a details/summary ?

File: site/content/docs/0.0/migration-from-boosted.md [hunk 6]
```diff
@@ -55,6 +55,32 @@ Technically, it means that you can get rid of the following things:
...
 
+### Border
+
+- <span class="badge text-bg-danger">Breaking</span> Border operative utilities have been removed: `.border-0`, `.border-top-0`, `.border-bottom-0`, `.border-start-0`, and `.border-end-0`. Please check the new [border values]({{< docsref "/utilities/borders#border" >}}) directly in the documentation and adapt your websites to them. You can still have them using `$enable-bootstrap-compatibility`.
+
+- <span class="badge text-bg-success">New</span> Border operative utilities: `.border-none`, `.border-top-none`, `.border-bottom-none`, `.border-start-none`, and `.border-end-none`.
+
+- <span class="badge text-bg-danger">Breaking</span> Border width utilities have been removed: `.border-0`, `.border-1`, `.border-2`, `.border-3`, `.border-4` and `.border-5`. Please check the new [border values]({{< docsref "/utilities/borders#width" >}}) directly in the documentation and adapt your websites to them. You can still have them using `$enable-bootstrap-compatibility`.
+
+- <span class="badge text-bg-success">New</span> Border width utilities: `.border-default`, `.border-none`, `.border-thin`, `.border-thick`, `.border-thicker` and `.border-thickest`.
```
Comment (vprothais) [inline site/content/docs/0.0/migration-from-boosted.md:66]: `.border-default` should be removed here

File: site/content/docs/0.0/utilities/borders.md (hunk 1 above)
Comment (vprothais) [reply site/content/docs/0.0/utilities/borders.md:211]: We still have the problem in compatibility mode, the rounded-2 is valued to 0, but rounded-1 have a non zero value. We could modify the map in utilities, in order to remove the usage of `var(--#{$prefix}border-radius)` for the size 2 or we could change the value of this CSS variables in root for another value than the default one.

I find it more logical to keep the default value in the CSS variable and to change the variables used in the map. But in every case we could have some issue with other libs using those variables.

File: .stylelintignore (hunk 4 above)
Comment (vprothais) [reply .stylelintignore:2]: I just saw that in [_utilities.test.scss](https://github.com/Orange-OpenSource/Orange-Boosted-Bootstrap/blob/ouds/main/scss/tests/mixins/_utilities.test.scss#L267) we already have generated content with leading 0, and the linter has been disabled only for the concerned line. Maybe we should stick to this in order to be consistent ? I'll do that until we decide what we choose to do.

File: site/content/docs/0.0/migration.md [hunk 7]
```diff
@@ -17,14 +17,31 @@ From now on, OUDS Web won't embed Bootstrap elements that are not part of Orange
 
 ### Utilities
 
+#### Border
+
+- <span class="badge text-bg-success">New</span> Border operative utilities: `.border`, `.border-none`, `.border-top`, `.border-top-none`, `.border-bottom`, `.border-bottom-none`, `.border-start`, `.border-start-none`, `.border-end` and `.border-end-none`.
```
[the remark this answers is not in the document]
Comment (vprothais) [reply site/content/docs/0.0/migration.md:18]: I resolve the conversation. The utilities are all new compared to previous version (we should not compare to boosted here)

File: .stylelintignore (hunk 4 above)
Comment (vprothais) [reply .stylelintignore]: We talked about this with @julien-deramond and decided to just disable the rule for the corresponding lines with a comment, like it's done in utilities test file.

File: site/content/docs/0.0/utilities/borders.md (hunk 1 above)
Comment (vprothais) [reply site/content/docs/0.0/utilities/borders.md:215]: We decided to keep the default token for the `rounded` helper, meaning it does nothing in Orange. We added a info callout to explain why. The mapping for `rounded-*` size in compatibility mode have been changed in order to remove `$border-radius` from this list for size 2. We introduced a arbitrary new value, without any variable for the size 5 (should be equivalent to an hypothetical xxxl)

File: site/content/docs/0.0/utilities/borders.md [hunk 8]
```diff
@@ -8,4 +8,315 @@ aliases:
...
+<hr class="text-body-secondary">
+<hr>
+{{< /example >}}-->
+
+## Drag and drop
+
+Use the `.border-drag` utility to apply the border style for a drop zone.
+
+{{< callout danger >}}
+This utility must not be used in any other context than drag and drop.
```
Comment (julien-deramond) [inline site/content/docs/0.0/utilities/borders.md:249]: Like in our docs 🙃? 

Screenshot 2024-09-16 at 08 26 13

Just kidding, we need to check with the designers how to render this section. I liked the fact that the border is a little bit different showing that it's not mandatory to open/read.

**Review by julien-deramond — COMMENTED** [review]
I haven't finished the review yet, here are some first comments

File: site/content/docs/0.0/utilities/borders.md (hunk 8 above)
Comment (vprothais) [reply site/content/docs/0.0/utilities/borders.md:249]: ahah, do as I say, not as I do :)
You're right, we should ask designer on how to render this kind of content.

File: scss/_utilities.scss
[the remark this answers is not in the document]
Comment (vprothais) [reply scss/_utilities.scss:1]: Do we add them to `*-ouds` keys too, or just for entries/modification on original utilities ?

[the remark this answers is not in the document]
Comment (julien-deramond) [reply scss/_utilities.scss:1]: Basically, everything that is different from the original Bootstrap file was tagged with `Boosted mod`. It's just helpful while integrating Bootstrap commits in the project. So all new stuff, all modified stuff with the "instead of" sentence if possible, all deleted stuff with the "no xxxx" comment.

File: site/content/docs/0.0/utilities/borders.md [hunk 9]
```diff
@@ -8,4 +8,315 @@ aliases:
...
+
+{{< scss-docs name="root-border-var" file="scss/_root.scss" >}}
+
+{{< /bootstrap-compatibility >}}
+
+### Sass tokens
+
+#### Raw tokens
+
+Border raw tokens<!-- generated via StyleDictionary--> as Sass variables. **Not to be used as-is**.
```
Comment (julien-deramond) [inline site/content/docs/0.0/utilities/borders.md:272]: See https://github.com/Orange-OpenSource/Orange-Boosted-Bootstrap/pull/2687/files#r1760598205

File: site/content/docs/0.0/utilities/opacity.md [hunk 10]
```diff
@@ -60,15 +60,17 @@ Set the `opacity` of an element using `.opacity-{value}` utilities.
 
 ### Sass tokens
 
-Opacity tokens as Sass variables.
+#### Raw tokens
 
-{{< scss-docs name="ouds-opacities-variables" file="scss/tokens/_raw.scss" >}}
+Opacity raw tokens<!-- generated via StyleDictionary--> as Sass variables. **Not to be used as-is**.
```
[the remark this answers is not in the document]
Comment (julien-deramond) [reply site/content/docs/0.0/utilities/opacity.md]: OK, let's tackle "Not to be used as-is." in another PR.
The other comment in "borders.md" file has been removed via https://github.com/Orange-OpenSource/Orange-Boosted-Bootstrap/pull/2687/commits/450e3f0d13d8efc44d95f8c01fbcd44ff278fb75

File: site/content/docs/0.0/utilities/opacity.md [hunk 11]
```diff
@@ -60,15 +60,17 @@ Set the `opacity` of an element using `.opacity-{value}` utilities.
...
 
-### Sass maps
+{{< scss-docs name="ouds-raw-opacity" file="scss/tokens/_raw.scss" >}}
 
-Opacity utilities are declared via a Sass map and then generated with our utilities API.
+#### Semantic tokens
 
-{{< scss-docs name="ouds-opacities-variables-maps" file="scss/tokens/_semantic.scss" >}}
+Opacity semantic tokens as Sass map. Opacity utilities are declared via a Sass map and then generated with our utilities API.
```
[the remark this answers is not in the document]
Comment (julien-deramond) [reply site/content/docs/0.0/utilities/opacity.md]: Applied to border page as well via https://github.com/Orange-OpenSource/Orange-Boosted-Bootstrap/pull/2687/commits/176b7fb13eb875cd63d84fd241ffef1213b7c3e0

File: scss/tokens/_raw.scss [hunk 12]
```diff
@@ -16,4 +16,35 @@ $ouds-opacity-500: .32 !default;
...
+$ouds-border-radius-75:   $ouds-border-base * .75 !default;
+// $ouds-border-radius-100:  $ouds-border-base !default;
+$ouds-border-radius-150:  $ouds-border-base * 1.5 !default;
+// $ouds-border-radius-200:  $ouds-border-base * 2 !default;
+$ouds-border-radius-300:  $ouds-border-base * 3 !default;
+// $ouds-border-radius-400:  $ouds-border-base * 4 !default;
+// $ouds-border-radius-500:  $ouds-border-base * 5 !default;
+// $ouds-border-radius-600:  $ouds-border-base * 6 !default;
+// $ouds-border-radius-800:  $ouds-border-base * 8 !default;
+
```
Comment (julien-deramond) [inline scss/tokens/_raw.scss:45]: I don't remember precisely what we said, but there's still a design token like this in our references outside of this PR. Let's check if we need to update the reference of this PR:
* `border-radius-9999` → `2000`

Comment (julien-deramond) [reply scss/tokens/_raw.scss:45]: OK, decision taken in https://github.com/Orange-OpenSource/Orange-Boosted-Bootstrap/pull/2687#discussion_r1740871202. I resolve this comment.

File: scss/tests/customize/_ouds-web-bootstrap.test.scss
Comment (julien-deramond) [inline scss/tests/customize/_ouds-web-bootstrap.test.scss:1]: Note: rounded utilities must be added to this unit test too

Comment (julien-deramond) [reply scss/tests/customize/_ouds-web-bootstrap.test.scss:1]: Done via https://github.com/Orange-OpenSource/Orange-Boosted-Bootstrap/pull/2687/commits/265a2d29dcd78fb93f322421a0ac15daf9f60a4c
It was a nightmare 😄 We need to find a way so it's easier in the future 🚀

File: site/content/docs/0.0/migration.md [hunk 13]
```diff
@@ -19,12 +40,12 @@ From now on, OUDS Web won't embed Bootstrap elements that are not part of Orange
...
 #### Opacity
 
-- <span class="badge text-bg-success">New</span> Opacity utilities: `.opacity-transparent`, `.opacity-weaker`, `.opacity-weak`, `.opacity-medium`, `.opacity-emphasis` and `.opacity-opaque`.
+- <span class="badge text-bg-success">New</span> Opacity utilities: `.opacity-transparent`, `.opacity-weaker`, `.opacity-weak`, `.opacity-medium`, `.opacity-strong` and `.opacity-opaque`.
 
 ### CSS and Sass variables
 
 - <span class="badge text-bg-success">New</span> `$enable-bootstrap-compatibility` option set to `false` by default. This option allows you to compile the Sass files with a Bootstrap compatibility mode.
-  - For instance, without the Bootstrap compatibility mode, you won't have the opacity Bootstrap utilities such as `opacity-0`, `opacity-1`, `opacity-50`, etc. You will only have the semantic OUDS Web utilities such as `opacity-transparent`, `opacity-weaker`, `opacity-weak`, etc.
+  - For instance, without the Bootstrap compatibility mode, you won't have the opacity Bootstrap utilities using numbers. You will only have the semantic OUDS Web utilities words instead.
```
Comment (julien-deramond) [inline site/content/docs/0.0/migration.md:48]: @vprothais Would you be OK if we revert this change? @hannahiss and I kinda prefer the old one with examples, that seems easier to understand on our side.

Comment (vprothais) [reply site/content/docs/0.0/migration.md:48]: That's good for me too, I also find the previous one more understandable :)

**Review by julien-deramond — APPROVED** [review]
Helped to finalize the PR. LGTM! 🚀 
Good job all!
